### ISSUE TITLE
ib.reqTickByTickData raising exception: 'ConnectionError: Socket disconnect'

### ISSUE THREAD

**Issue by miali88** [erdewit/ib_insync#480]
I am attempting to place limit orders based on the prevailing best bid/offer. I am attempting to do this through 'ib.reqTickByTickData(symbol, 'BidAsk')'. This function works perfectly when testing. However when I put it into a while True loop, it acts up and causes the following error: 

```
CancelledError                           Traceback (most recent call last)
File c:\Users\Azkar\miniconda3\envs\quant\lib\asyncio\tasks.py:234, in Task.__step(***failed resolving arguments***)
    [233](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=232)     else:
--> [234](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=233)         result = coro.throw(exc)
    [235](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=234) except StopIteration as exc:

File c:\Users\Azkar\miniconda3\envs\quant\lib\asyncio\tasks.py:605, in sleep(delay, result)
    [604](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=603) try:
--> [605](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=604)     return await future
    [606](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=605) finally:

File c:\Users\Azkar\miniconda3\envs\quant\lib\asyncio\futures.py:284, in Future.__await__(self)
    [283](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=282)     self._asyncio_future_blocking = True
--> [284](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=283)     yield self  # This tells Task to wait for completion.
    [285](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=284) if not self.done():

File c:\Users\Azkar\miniconda3\envs\quant\lib\asyncio\tasks.py:304, in Task.__wakeup(self, future)
    [303](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=302) try:
--> [304](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=303)     future.result()
    [305](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=304) except BaseException as exc:
    [306](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/tasks.py?line=305)     # This may also be a cancellation.

File c:\Users\Azkar\miniconda3\envs\quant\lib\asyncio\futures.py:196, in Future.result(self)
    [195](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=194)     exc = self._make_cancelled_error()
--> [196](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=195)     raise exc
    [197](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=196) if self._state != _FINISHED:

CancelledError: 

During handling of the above exception, another exception occurred:

CancelledError                            Traceback (most recent call last)
File c:\Users\Azkar\miniconda3\envs\quant\lib\site-packages\ib_insync\util.py:332, in run(timeout, *awaitables)
    [331](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=330) try:
--> [332](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=331)     result = loop.run_until_complete(task)
    [333](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=332) except asyncio.CancelledError as e:

File c:\Users\Azkar\miniconda3\envs\quant\lib\site-packages\nest_asyncio.py:70, in _patch_loop.<locals>.run_until_complete(self, future)
     [68](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/nest_asyncio.py?line=67)     raise RuntimeError(
     [69](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/nest_asyncio.py?line=68)         'Event loop stopped before Future completed.')
---> [70](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/nest_asyncio.py?line=69) return f.result()

File c:\Users\Azkar\miniconda3\envs\quant\lib\asyncio\futures.py:196, in Future.result(self)
    [195](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=194)     exc = self._make_cancelled_error()
--> [196](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=195)     raise exc
    [197](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/asyncio/futures.py?line=196) if self._state != _FINISHED:

CancelledError: 

During handling of the above exception, another exception occurred:

ConnectionError                           Traceback (most recent call last)
c:\Users\Azkar\OneDrive\Programming\poseiden_engine\MA_trend\MA_IB_live_algo\limit_multi.ipynb Cell 5' in <cell line: 32>()
     [33](vscode-notebook-cell:/c%3A/Users/Azkar/OneDrive/Programming/poseiden_engine/MA_trend/MA_IB_live_algo/limit_multi.ipynb#ch0000020?line=32) for symbol in contracts:
     [34](vscode-notebook-cell:/c%3A/Users/Azkar/OneDrive/Programming/poseiden_engine/MA_trend/MA_IB_live_algo/limit_multi.ipynb#ch0000020?line=33)     req_real = ib.reqTickByTickData(symbol, 'BidAsk')
---> [35](vscode-notebook-cell:/c%3A/Users/Azkar/OneDrive/Programming/poseiden_engine/MA_trend/MA_IB_live_algo/limit_multi.ipynb#ch0000020?line=34)     ib.sleep(1)
     [36](vscode-notebook-cell:/c%3A/Users/Azkar/OneDrive/Programming/poseiden_engine/MA_trend/MA_IB_live_algo/limit_multi.ipynb#ch0000020?line=35)     place_order('BUY', 1, symbol, req_real.ask)
     [37](vscode-notebook-cell:/c%3A/Users/Azkar/OneDrive/Programming/poseiden_engine/MA_trend/MA_IB_live_algo/limit_multi.ipynb#ch0000020?line=36)     ib.cancelTickByTickData(symbol,'BidAsk')

File c:\Users\Azkar\miniconda3\envs\quant\lib\site-packages\ib_insync\util.py:378, in sleep(secs)
    [370](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=369) def sleep(secs: float = 0.02) -> bool:
    [371](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=370)     """
    [372](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=371)     Wait for the given amount of seconds while everything still keeps
    [373](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=372)     processing in the background. Never use time.sleep().
   (...)
    [376](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=375)         secs (float): Time in seconds to wait.
    [377](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=376)     """
--> [378](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=377)     run(asyncio.sleep(secs))
    [379](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=378)     return True

File c:\Users\Azkar\miniconda3\envs\quant\lib\site-packages\ib_insync\util.py:334, in run(timeout, *awaitables)
    [332](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=331)     result = loop.run_until_complete(task)
    [333](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=332) except asyncio.CancelledError as e:
--> [334](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=333)     raise globalErrorEvent.value() or e
    [335](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=334) finally:
    [336](file:///c%3A/Users/Azkar/miniconda3/envs/quant/lib/site-packages/ib_insync/util.py?line=335)     globalErrorEvent.disconnect(onError)

ConnectionError: Socket disconnect
```

The code simplified as per below:

```
import pandas as pd
import numpy as np
from datetime import datetime
from ib_insync import *
util.startLoop()
util.logToConsole()

ib = IB()
ib.connect('127.0.0.1', 6000, clientId=1)

contracts = [Stock(stonk,'SMART','USD') for stonk in ('AAPL','TSLA','JPM')]
ib.qualifyContracts(*contracts)

for contract in contracts:
    bar = ib.reqHistoricalData(contract, 
                             endDateTime='', 
                             durationStr='10 D',
                             barSizeSetting='10 mins', 
                             whatToShow='TRADES', 
                             useRTH=True,
                             keepUpToDate=True)

def place_order(direction, size, contract, price):
    if direction == 'BUY': 
        contract_order = LimitOrder('BUY', size, price)  
        trade = ib.placeOrder(contract, contract_order)

    elif direction == 'SELL':
        contract_order = LimitOrder('SELL', size, price)         
        trade = ib.placeOrder(contract, contract_order) 

while True:
    for symbol in contracts:
        req_real = ib.reqTickByTickData(symbol, 'BidAsk')
        ib.sleep(1)
        place_order('BUY', 1, symbol, req_real.ask)
        ib.cancelTickByTickData(symbol,'BidAsk')
```

Tried different variations of the code but the error persists. Thinking of using another way to get current bid/ask if this doesn't work. 

Cheers

**Comment by mrwillett** [comment]
Firstly, I'm genuinely curious why you would event want to place **any** kind of order in an infinite loop unless you're trying to set a record for fastest margin call -- although in reality IB just liquidates your assets without the courtesy.

Regardless, I wan't able to replicate your error. I did have to add a check for a valid `req_real.ask` though this is tangential to your reported bug and I presume you're doing this and more in the actual code.

> Tried different variations of the code but the error persists

Interesting. I get no errors with:

```py
from ib_insync import IB, Index, util, Stock, LimitOrder


def place_order(direction, size, contract, price):
    print(f"placing order for price: {price}")
    if direction == 'BUY': 
        contract_order = LimitOrder('BUY', size, price)  
        trade = ib.placeOrder(contract, contract_order)

    elif direction == 'SELL':
        contract_order = LimitOrder('SELL', size, price)         
        trade = ib.placeOrder(contract, contract_order)



def main():
    ib.connect('127.0.0.1', 7497, clientId=0)
    ib.sleep(1)
    print(f"{'connected' if ib.isConnected() else 'connection failed'}")
    # --------------------------------

    contracts = [Stock(stonk,'SMART','USD') for stonk in ('AAPL','TSLA','JPM')]
    ib.qualifyContracts(*contracts)

    while True:
        for symbol in contracts:
            req_real = ib.reqTickByTickData(symbol, 'BidAsk')
            while (util.isNan(req_real.ask)):
                ib.sleep(0.1)
            print(f"req_real: {req_real}")
            ib.sleep(1)
            place_order('BUY', 1, symbol, req_real.ask)
            print("about to cancel tick data...")
            ib.cancelTickByTickData(symbol,'BidAsk')

    # --------------------------------
    ib.sleep(3)
    ib.disconnect()


if __name__ == "__main__":
    ib = IB()
    main()
```

> Thinking of using another way to get current bid/ask if this doesn't work.

I think a better architecture would be to request all the data lines outside the loop, wait for values, the enter your while(1) loop, otherwise you will have a lot of overhead from just requesting and cancelling data lines. Not to mention, according to [ibkr:](https://interactivebrokers.github.io/tws-api/tick_data.html)
> No more than 1 tick-by-tick request can be made for the same instrument within 15 seconds

Here's a crude example:

```py
tick_bars = [ib.reqTickByTickData(contract, 'BidAsk') for contract in contracts]  # request data here once

while True:  # def margin_call_pls()?
    for ticker in tick_bars:
        while (util.isNan(ticker.ask)):
            ib.sleep(0.1)
        print(f"symbol: {ticker.contract.symbol}, ask: {ticker.ask}")
        # ib.sleep(1)  # purely to slow execution since a (still potentially invalid) real number is guaranteed by isNan()
        place_order('BUY', 1, ticker.contract, ticker.ask)
        print("about to cancel tick data...")
    
for ticker.contract.symbol in tick_bars:  # *now* we cancel the data
ib.cancelTickByTickData(symbol, 'BidAsk')
```

the output is a terrifying number of orders:
```
symbol: AAPL, ask: 137.3
placing order for price: 137.3
about to cancel tick data...
symbol: TSLA, ask: 704.3
placing order for price: 704.3
about to cancel tick data...
symbol: JPM, ask: 119.55
placing order for price: 119.55
about to cancel tick data...
symbol: AAPL, ask: 137.3
placing order for price: 137.3
about to cancel tick data...
.
.
.
```

**Comment by miali88** [comment]
Thanks for your response @mrwillett. 

I can assure I'm not interested in margin calls. The code above was a lazy extract from an existing code for illustration purposes. The place order function is only called when certain conditions are met inside the while true loop. 

I like your idea of placing the reqTick data outside of the loop. Unfortunately I still get the socket disconnect error. And for the first suggestion nothing happens, the script just prints 'connected'. 

I'm finding the reqHistoricalTicks function to work just fine. I set the number of ticks to just 1, and set the bid/ask around that.

**Comment by mrwillett** [comment]
Interesting. So it seems like it's something in your environment. Your traceback suggests something not playing nicely with asyncio. Can you provide your `quant` enviroment packages with `pip freeze > requirements.txt` as well as the version of python you are running? I also notice you're runnign in a notebook environment. Try running just a plain python script (I'm assuming this is how you ran the code snippet I provided above). I'm suspicious that this has to do with asyncio and util.startLoop().

When you say the code from the first suggestion only prints "connected" (but doesn't give a socket disconnect) this is actually normal behavior. Nothing being printed means that you are still waiting for the first ask price to be not `nan`. Try running the below (during market hours) as a python script. Even with good internet during market hours for the given instrument, it took up to 30 seconds for me to get a valid ask price.

Try running this as a python script and report the output:
```py
from ib_insync import IB, Index, util, Stock, LimitOrder
# importantly, not using util.startLoop()! (and not running an ipynb environment)


def place_order(direction, size, contract, price):
    print(f"placing order for price: {price}")
    # * actual order placement code not needed for
    #   this example since this is not the issue


def main():
    ib.connect('127.0.0.1', 6000, clientId=0)
    ib.sleep(1)
    print(f"{'connected' if ib.isConnected() else 'connection failed'}")
    # --------------------------------

    contracts = [Stock(stonk,'SMART','USD') for stonk in ('AAPL','TSLA','JPM')]
    ib.qualifyContracts(*contracts)

    tick_bars = [ib.reqTickByTickData(contract, 'BidAsk') for contract in contracts]
    ran = False
    try:
        while True:
            for ticker in tick_bars:
                while (util.isNan(ticker.ask)):
                    if not ran:
                        print(f"waiting on a valid ask price for {ticker.contract.symbol}")
                        ran = True
                    ib.sleep(0.1)
                ran = False
                print(f"symbol: {ticker.contract.symbol}, ask: {ticker.ask}")
                place_order('BUY', 1, ticker.contract, ticker.ask)
    except KeyboardInterrupt:
        # `ctrl-C` to be able to reach the below and properly cancel the data
        for ticker in tick_bars:
            ib.cancelTickByTickData(ticker.contract.symbol, 'BidAsk')
        ib.sleep(3)
        ib.disconnect()



if __name__ == "__main__":
    ib = IB()
    main()
```

If you still get a socket disconnect please provide some more verbose logging of when exactly execution stops. Either a log file or a bunch of `print()` statements with unique identifiers so I can see what is being called and when before the error would help significantly.

**Comment by 5KevinH** [comment]
Sorry for intruding on the conversation. I am now following this matter with interest.
@miali88 - Does this problem occur when connecting via TWS, IB Gateway, or both?
Thanks.

**Comment by 5KevinH** [comment]
@mrwillett - my problem seems to have been solved by referring to your solution immediately above.
Thanks,
KH

**Comment by mrwillett** [comment]
@5KevinH If you had the same issue as @miali88 and it was solved using the above, plus me not being able to replicate the error, this issue can be closed.

closes #480

**Comment by 5KevinH** [comment]
Thanks @mrwillett I didn't have exactly the same problem, but one very similar. I do suspect, however, that the problem @miali88 will be fixed by your code sample.
One thing, if I may. cancelTickByTickData is looking for exactly the same contract as was provided by reqTickByTickData. So, in your solution, I would change this:
`        for ticker in tick_bars:
            ib.cancelTickByTickData(ticker.contract.symbol, 'BidAsk')`
to this:
`        for contract in contracts:
            ib.cancelTickByTickData(contract, 'BidAsk')`
Thanks again for your help.
KH

**Comment by erdewit** [comment]
I can't reproduce the issue either with this simplified case:

```python
contracts = [Stock(symbol,'SMART','USD') for symbol in ('AAPL','TSLA','JPM')]

for i in range(10):
    tickers = [ib.reqTickByTickData(contract, 'BidAsk') for contract in contracts]
    ib.sleep(1)
    print(tickers)
    for contract in contracts:
        ib.cancelTickByTickData(contract, 'BidAsk')
```

If the connection is actively dropped by TWS/gateway then that tends to be a sort of punishment for what the API is doing.
